### DISCUSSION TITLE
Update ui.table and ui.aggrid after modifying the underlying pandas dataframe

### DISCUSSION THREAD

**Discussion by rohitsathish** [zauberzeug/nicegui#2731] [Ideas / Feature Requests]
### Question

It looks like ```.update``` does not work when the source dataframe of ```ui.table.from_pandas``` or ```ui.aggrid.from_pandas``` is changed. I even tried with the ```app.storage``` approach and that doesn't work either. Is there an alternative approach that I'm missing?

```python
from nicegui import ui, app
import pandas as pd
    
@ui.page('/')
async def main_ui():
    controls = ui.row()

    df = pd.DataFrame({'a': [1,2], 'b': [3,4]})
    app.storage.user['df'] = df.to_dict()

    df_ui = ui.aggrid.from_pandas(df).classes('max-h-24')
    df_ui1 = ui.table.from_pandas(df)

    df_ui3 = ui.aggrid.from_pandas(pd.DataFrame(app.storage.user['df'])).classes('max-h-24')
    df_ui4 = ui.table.from_pandas(pd.DataFrame(app.storage.user['df']))

    async def df_update():
        nonlocal df
        df = df.iloc[0]
        app.storage.user['df'] = pd.DataFrame(app.storage.user['df']).iloc[0].to_dict()
        print(df), print(app.storage.user['df'])
        df_ui.update(), df_ui1.update(), df_ui3.update(), df_ui4.update()

    with controls:
        ui.button('Update', on_click=df_update)

ui.run(port=8082, storage_secret='test')
```

**Comment by falkoschindler** [comment]
Hi @rohitsathish,

This is currently not implemented. The `from_pandas` methods are only for creating new `ui.table` or `ui.aggrid` instances. We could add something like `update_from_pandas` though. I'll turn this discussion into a feature request.

**Comment by tmlmt** [comment]
Hi guys, I stumbled upon the same issue and would love this feature request to be worked on.
Even reloading the page doesn't force the update of the table, so it's rather annoying. 
In the meantime, how can we update the tables initially generated with from_pandas? 
- Shall we revert to defining tables manually with columns and rows? 
- or shall we regenerate with from_pandas, and remove the old table and add the new one into the corresponding container?

**Reply by falkoschindler** [reply]
I guess you can take the source code and modify it to update an existing table (or AG Grid):
https://github.com/zauberzeug/nicegui/blob/7a3cfc2c9ed1f6d1532f62db061cd2451861b3b5/nicegui/elements/table.py#L89-L135
Apart from that, the best way to push this feature forward would be to create a pull request. We would be happy to review and add it to the next release. 🙂

**Reply by tmlmt** [reply]
I will give it a try 😊

In the meantime, as a workaround, I've used a container that I clear  and regenerate from_pandas:

```python
def update_table() -> None:
    table_container.clear()
    with table_container:
        ui.table.from_pandas(df)

<...>

with ui.element() as table_container:
    ui.table.from_pandas(df)

```

**Comment by falkoschindler** [comment]
The `update_from_pandas()` method has been implemented in #3525 and will be released shortly in version 2.0.0.
